src/Data/ByteString/Base32/Internal/Loop.hs

```diff
@@ -88,48 +89,120 @@ decodeLoop !lut !dptr !sptr !end finish = go dptr sptr 0
-            !b = lix (unsafeShiftR t 48)
-            !c = lix (unsafeShiftR t 40)
-            !d = lix (unsafeShiftR t 32)
-            !e = lix (unsafeShiftR t 24)
-            !f = lix (unsafeShiftR t 16)
-            !g = lix (unsafeShiftR t 8)
-            !h = lix t
-
-        if
-          | a == 0x63 -> padErr src
-          | b == 0x63 -> padErr (plusPtr src 1)
-          | c == 0x63 -> padErr (plusPtr src 2)
-          | d == 0x63 -> padErr (plusPtr src 3)
-          | e == 0x63 -> padErr (plusPtr src 4)
-          | f == 0x63 -> padErr (plusPtr src 5)
-          | g == 0x63 -> padErr (plusPtr src 6)
-          | h == 0x63 -> padErr (plusPtr src 7)
-          | a == 0xff -> err src
-          | b == 0xff -> err (plusPtr src 1)
-          | c == 0xff -> err (plusPtr src 2)
-          | d == 0xff -> err (plusPtr src 3)
-          | e == 0xff -> err (plusPtr src 4)
-          | f == 0xff -> err (plusPtr src 5)
-          | g == 0xff -> err (plusPtr src 6)
-          | h == 0xff -> err (plusPtr src 7)
-          | otherwise -> do
-
-            let !w = ((unsafeShiftL a 35)
-                  .|. (unsafeShiftL b 30)
-                  .|. (unsafeShiftL c 25)
-                  .|. (unsafeShiftL d 20)
-                  .|. (unsafeShiftL e 15)
-                  .|. (unsafeShiftL f 10)
-                  .|. (unsafeShiftL g 5)
-                  .|. h) :: Word64
-
-            poke @Word32 (castPtr dst) (byteSwap32 (fromIntegral (unsafeShiftR w 8)))
-            poke @Word8 (plusPtr dst 4) (fromIntegral w)
-            go (plusPtr dst 5) (plusPtr src 8) (n + 5)
+        let a = lix (unsafeShiftR t 56)
+            b = lix (unsafeShiftR t 48)
+            c = lix (unsafeShiftR t 40)
+            d = lix (unsafeShiftR t 32)
+            e = lix (unsafeShiftR t 24)
+            f = lix (unsafeShiftR t 16)
+            g = lix (unsafeShiftR t 8)
+            h = lix t
+
+        decodeChunk dst src a b c d e f g h
+
+    finalChunk !dst !src !a !b !c !d !e !f !g !h
+      | a == 0x63 = padErr src
+      | b == 0x63 = padErr (plusPtr src 1)
+      | a == 0xff = err src
+      | b == 0xff = err (plusPtr src 1)
+      | c == 0xff = err (plusPtr src 2)
+      | d == 0xff = err (plusPtr src 3)
+      | e == 0xff = err (plusPtr src 4)
+      | f == 0xff = err (plusPtr src 5)
+      | g == 0xff = err (plusPtr src 6)
+      | h == 0xff = err (plusPtr src 7)
+      | otherwise = do
+
+        let !o1 = (fromIntegral a `unsafeShiftL` 3) .|. (fromIntegral b `unsafeShiftR` 2)
+            !o2 = (fromIntegral b `unsafeShiftL` 6)
+              .|. (fromIntegral c `unsafeShiftL` 1)
+              .|. (fromIntegral d `unsafeShiftR` 4)
+            !o3 = (fromIntegral d `unsafeShiftL` 4) .|. (fromIntegral e `unsafeShiftR` 1)
+            !o4 = (fromIntegral e `unsafeShiftL` 7)
+              .|. (fromIntegral f `unsafeShiftL` 2)
+              .|. (fromIntegral g `unsafeShiftR` 3)
+            !o5 = (fromIntegral g `unsafeShiftL` 5) .|. fromIntegral h
```

Comment:
We must explicitly split up our final octet here so as not to leak bits into the slots determined by the pads.